src/python/pants/backend/adhoc/adhoc_tool_test.py

```diff
@@ -374,3 +374,60 @@ def test_adhoc_tool_workspace_invalidation_sources(rule_runner: PythonRuleRunner
     (Path(rule_runner.build_root) / "src" / "a-file").write_text("xyzzy")
     result3 = execute_adhoc_tool(rule_runner, address)
     assert result1.snapshot != result3.snapshot
+
+
+def test_adhoc_tool_path_env_modify_mode(rule_runner: PythonRuleRunner) -> None:
+    expected_path = "/bin:/usr/bin"
+    rule_runner.write_files(
+        {
+            "src/BUILD": dedent(
+                f"""\
+            system_binary(name="bash", binary_name="bash")
+            system_binary(name="renamed_cat", binary_name="cat")
+            adhoc_tool(
+                name="shims_prepend",
+                runnable=":bash",
+                args=["-c", "echo $PATH > foo.txt && renamed_cat foo.txt > path.txt"],
+                extra_env_vars=["PATH={expected_path}"],
+                output_files=["path.txt"],
+                runnable_dependencies=[":renamed_cat"],
+                path_env_modify="prepend",
+            )
+            adhoc_tool(
+                name="shims_append",
+                runnable=":bash",
+                args=["-c", "echo $PATH > foo.txt && renamed_cat foo.txt > path.txt"],
+                extra_env_vars=["PATH={expected_path}"],
+                output_files=["path.txt"],
+                runnable_dependencies=[":renamed_cat"],
+                path_env_modify="append",
+            )
+            adhoc_tool(
+                name="shims_off",
+                runnable=":bash",
+                args=["-c", "echo $PATH > foo.txt && renamed_cat foo.txt > path.txt"],
+                extra_env_vars=["PATH={expected_path}"],
+                output_files=["path.txt"],
+                runnable_dependencies=[":renamed_cat"],
+                path_env_modify="off",
+            )
+            """
+            )
+        }
+    )
+
+    def run(target_name: str) -> str:
+        result = execute_adhoc_tool(rule_runner, Address("src", target_name=target_name))
+        contents = rule_runner.request(DigestContents, [result.snapshot.digest])
+        assert len(contents) == 1
+        return contents[0].content.decode().strip()
+
+    path_prepend = run("shims_prepend")
+    assert path_prepend.endswith(expected_path)
+    assert len(path_prepend) > len(expected_path)
+
+    path_append = run("shims_append")
+    assert path_append.startswith(expected_path)
+    assert len(path_append) > len(expected_path)
+
+    # TODO: shims_off will fail due to `renamed_cat` not being on the PATH. How do we obtain FailableProcessResult here?
```

Comment:
This should probably be resolved before merging. It sounds like a bug.

Reply:
Will add a test, but it should not be bug: When path modifications are turned off, the binary shims directory would not be on the PATH and so `renamed_cat` would be expected to not be found.

Reply:
The "How do we obtain FailableProcessResult here?" question relates to how can I actually write the test. I need to test for the failure and the engine API used in this file does not have a version which does not through an exception. I could capture the exception and check the message, but that is not my first choice of how to write the test.

Reply:
Ah. That makes more sense.

Reply:
I came up with a way to test it without needing access to the failed result without an exception.